src/backend/timelineCredPlan.test.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+
+import {TestTaskReporter} from "../util/taskReporter";
+import {timelineCredPlan} from "./timelineCredPlan";
+
+const mockCompute = () => jest.fn();
+
+const fakeGraph = ({
+  toJSON: () => ({is: "fake-graph"}),
+}: any);
+
+const fakeCred = ({
+  toJSON: () => ({is: "fake-cred"}),
+}: any);
+
+describe("src/backend/timelineCredPlan", () => {
+  describe("timelineCredPlan", () => {
+    it("should defer to the provided compute function", async () => {
+      // Given
+      const plugins = [];
+      const reporter = new TestTaskReporter();
+      const params = {alpha: 0.456};
+      const compute = mockCompute();
+      compute.mockResolvedValueOnce(fakeCred);
+
+      // When
+      const computeCred = timelineCredPlan(plugins, reporter, compute);
+      const cred = await computeCred(fakeGraph, params);
+
+      // Then
+      expect(cred).toEqual(fakeCred);
+      expect(compute).toBeCalledTimes(1);
+      expect(compute).toBeCalledWith({
+        graph: fakeGraph,
+        params,
+        plugins,
+      });
+    });
+
+    it("should give the right tasks to the TaskReporter", async () => {
```

Comment:
If you wanna make this test more robust, you could verify that the task starts when compute starts, and task finishes after compute finishes. I.e. set up mock compute promise that will hang until you (synchronously) call a callback, and then:

- verify tasks are empty
- start computeCred
- verify there's one active task and one entry
- finish computeCred
- verify two entries and no active tasks

Reply:
I've not seen this before in the codebase, do we have an example for this?

To be honest though. I think the time and ordering accuracy of task reporter feedback doesn't warrant this type of unit tests. If with manual testing it provides enough feedback to know whether the program froze and/or which step it's at, that will suffice.

If we really care *that* much about whether a task starts and finishes at the right time, we should create a middleware for it and just test it once.

```js
function reportTask<T>(
	reporter: TaskReporter,
	name: string,
	yourTask: () => Promise<T>
): Promise<T> {
	reporter.start("name");
	const result = await yourTask();
	reporter.finish("name");
	return result;
}
```

This looks very familiar to the current implementation :smile: